### PR TITLE
add sentiment_analysis_cnn (#256)

### PR DESCRIPTION
**Pull request by astonzhang**
Merge @xiaotinghe 's PR:
https://github.com/dmlc/gluon-nlp/pull/256

## Description ##
Add sentiment analysis with textCNN including various data sets. This is intended to use convolutional neural networks (textCNN) for sentiment analysis.

- Convolutional neural networks for sentiment analysis
    - The script shows how to use convolutional neural networks (textCNN) for sentiment analysis on various datasets.

- Various datasets for sentiment analysis.
    - MR is a movie-review data set of 10,662 sentences labeled with respect to their overall sentiment polarity (positive or negative).
    - SST-1 is an extension of the MR data set. However, training/test splits are provided and labels are fine-grained (very positive, positive, neutral, negative, very negative). The training and test data sets have 237,107 and 2,210 sentences respectively.
    - SST-2 is the same with SST-1, but neutral sentences are removed and only binary sentiment polarity are considered: very positive is considered as positive, and very negative is considered as negative.
    - SUBJ is a Subjectivity data set for sentiment analysis. Sentences labeled with respect to their subjectivity status (subjective or objective).
    - TREC is a movie-review data set of 10,000 sentences labeled with respect to their subjectivity status (subjective or objective).

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [x] textCNN model.
- [x] textCNN running scripts.
- [x] Data sets (MR, SST, Subj, and TREC)

## Comments ##
- N.A.

### CODE REVIEW

File: gluonnlp/data/sentiment.py [hunk 1]
```diff
@@ -69,6 +70,203 @@ def _get_data(self):
...
+
+    Parameters
+    ----------
+    segment : str, default 'train'
+        Dataset segment. Options are 'train' and 'test'.
+    root : str, default '$MXNET_HOME/datasets/mr'
+        Path to temp folder for storing data.
+        MXNET_HOME defaults to '~/.mxnet'.
+    """
+    def __init__(self, segment='all', root=os.path.join(_get_home_dir(), 'datasets', 'mr')):
```
Comment (szha) [inline gluonnlp/data/sentiment.py]: seems that only "all" segment is supported. in that case, the segment argument can be removed.

**Comment by mli** [comment]
Job PR-391/2 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-391/2/index.html

File: gluonnlp/data/sentiment.py [hunk 2]
```diff
@@ -69,6 +70,202 @@ def _get_data(self):
...
+
+    def _get_data(self):
+        data_file_name, data_hash = self._data_file['all']
+        root = self._root
+        path = os.path.join(root, data_file_name)
+        if not os.path.exists(path) or not check_sha1(path, data_hash):
+            download(_get_repo_file_url('gluon/dataset/mr', data_file_name),
+                     path=root, sha1_hash=data_hash)
+
+    def _read_data(self):
```
Comment (szha) [inline gluonnlp/data/sentiment.py]: The `_read_data` and `_get_data` have opportunities for refactoring. Given the number of datasets that share the same logic, we may consider introducing a base class that provides `_get_data` and `_read_data`.

File: scripts/sentiment_analysis/sentiment_analysis_cnn.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+import text_cnn
+
+
+np.random.seed(3435)
+random.seed(3435)
+mx.random.seed(3435)
+
+parser = argparse.ArgumentParser(description='Sentiment analysis with the textCNN model on\
+                                 various datasets.')
+parser.add_argument('--data_name', choices=['MR', 'SST-1', 'SST-2', 'Subj', 'TREC'], default='MR',
```
Comment (sravanbabuiitm) [inline scripts/sentiment_analysis/sentiment_analysis_cnn.py]: This initialy is built for generic text classification usecase,wouldnt it be nice to take in any generic dataset as input and train a classifier from it ? rather than taking only predefined datasets ?

**Comment by mli** [comment]
Job PR-391/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-391/3/index.html

File: scripts/sentiment_analysis/process_data.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        train_dataset = nlp.data.SUBJ(root='data/Subj')
+        output_size = 2
+        return train_dataset, output_size
+    else:
+        train_dataset, test_dataset = [nlp.data.TREC(root='data/trec', segment=segment)
+                                       for segment in ('train', 'test')]
+        output_size = 6
+        return train_dataset, test_dataset, output_size
+
+def _clean_str(string, data_name):
```
Comment (sravanbabuiitm) [inline scripts/sentiment_analysis/process_data.py]: Can you add a comment above this method as to what the cleaning does for SST1/SST2 datasets and what cleaning is required for other datasets ?
like for ex: it replaces escape characters etc.

File: scripts/sentiment_analysis/process_data.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        line = train_dataset[i][0].split()
+        max_len = max_len if max_len > len(line) else len(line)
+        all_token.extend(line)
+    for i, line in enumerate(test_dataset):
+        test_dataset[i][0] = _clean_str(line[0], data_name)
+        line = test_dataset[i][0].split()
+        max_len = max_len if max_len > len(line) else len(line)
+    all_token.extend(line)
+    vocab = nlp.Vocab(nlp.data.count_tokens(all_token))
+    vocab.set_embedding(nlp.embedding.create('Word2Vec', source='GoogleNews-vectors-negative300'))
```
Comment (sravanbabuiitm) [inline scripts/sentiment_analysis/process_data.py]: Cant we take this too as input from customer ?
like w2v, Glove, fastText ?

in args ?

File: scripts/sentiment_analysis/process_data.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    for line in vocab.embedding._idx_to_token:
```
Comment (sravanbabuiitm) [inline scripts/sentiment_analysis/process_data.py]: [minor] rename line to word?

File: scripts/sentiment_analysis/process_data.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    lengths = gluon.data.SimpleDataset(pool.map(_get_length, dataset))
+    end = time.time()
+    print('Done! Tokenizing Time={:.2f}s, #Sentences={}'.format(end - start, len(dataset)))
+    return dataset, lengths
+
+def load_dataset(data_name):
+    """Load sentiment dataset."""
+    if data_name == 'MR' or data_name == 'Subj':
+        train_dataset, output_size = _load_file(data_name)
+        vocab, max_len = _build_vocab(data_name, train_dataset, [])
```
Comment (sravanbabuiitm) [inline scripts/sentiment_analysis/process_data.py]: Arent we truncating ?
like max(1000, np.percentile(lenghts,99)) etc just to safegaurd us ?

File: scripts/sentiment_analysis/sentiment_analysis_cnn.py (hunk 3 above)
Comment (astonzhang) [reply scripts/sentiment_analysis/sentiment_analysis_cnn.py]: Thanks. Due to the nature of scripts (rather than API), we usually dedicate scripts to certain tasks. Users can subclass `SimpleDataset` or tweak the scripts for other tasks. We may extend scripts to allow user-defined data sets as future work.

File: scripts/sentiment_analysis/process_data.py (hunk 7 above)
Comment (astonzhang) [reply scripts/sentiment_analysis/process_data.py]: These are according to experimental settings of the textCNN paper/implementations.

**Comment by mli** [comment]
Job PR-391/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-391/5/index.html

File: scripts/sentiment_analysis/process_data.py (hunk 4 above)
Comment (astonzhang) [reply scripts/sentiment_analysis/process_data.py]: Since re.sub has a clear documentation, it may be verbose to comment 10+ operations

File: scripts/sentiment_analysis/process_data.py (hunk 6 above)
Comment (astonzhang) [reply scripts/sentiment_analysis/process_data.py]: resolved

File: scripts/sentiment_analysis/process_data.py (hunk 5 above)
Comment (astonzhang) [reply scripts/sentiment_analysis/process_data.py]: Thanks. This script is dedicated to reproduce results of the textCNN paper, thus all the experimental settings are set the same as those in the paper. The model in the script hence assumes 300-dim word embeddings. Other embeddings may have varying dimensions such as 50, 100, 200 etc.

**Comment by astonzhang** [comment]
@xiaotinghe Please check my above changes. Let me know if you spot any issues. Again, thank you for your great contributions to gluonnlp :)

File: gluonnlp/data/sentiment.py (hunk 2 above)
Comment (astonzhang) [reply gluonnlp/data/sentiment.py]: resolved.

**Comment by mli** [comment]
Job PR-391/6 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-391/6/index.html

**Comment by mli** [comment]
Job PR-391/7 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-391/7/index.html

File: scripts/sentiment_analysis/sentiment_analysis_cnn.py (hunk 3 above)
Comment (szha) [reply scripts/sentiment_analysis/sentiment_analysis_cnn.py]: Regarding this request, once the scripts are in good quality, we may start vending it as part of the model zoo API as module functions (e.g. `python -m gluonnlp.model_zoo.sentiment_analysis --input-file 'imdb.txt'`). This is too large a scope to include in 0.5.0. Also, it will be a lot easier to push this forward once we standardize the scripts.

Comment (sravanbabuiitm) [reply scripts/sentiment_analysis/sentiment_analysis_cnn.py]: Thanks for clarification Sheng!

**Comment by astonzhang** [comment]
Thank @sravanbabuiitm and @szha for all the feedbacks!
Last thing: please keep commit history of @xiaotinghe when merging to master in order to acknowledge his contributions to gluonnlp.
